File: #41_First_Missing_Positive/Solution.py
'''
O(n) solution, O(1) space
1. be aware that max missing val is len(nums) + 1
2. use index information along with value itself
3. special case: swap with the same value
'''

class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def firstMissingPositive(self, nums):
        if not nums:
        	return 1
        for i in range(len(nums)):
        	while nums[i] != i + 1 and nums[i] > 0 and nums[i] <= len(nums) and nums[i] != nums[nums[i] - 1]:
        		x, y = i, nums[i] - 1
        		tmp = nums[x]
        		nums[x] = nums[y]
        		nums[y] = tmp
        for i, v in enumerate(nums):
        	if i + 1!= v:
        		return i + 1
        return len(nums) + 1

sol = Solution()
print sol.firstMissingPositive([1, 1])

File: #41_First_Missing_Positive/Solution.java
public class Solution {
    public int firstMissingPositive(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
        	while (nums[i] > 0 && nums[i] <= nums.length) {
        		if (nums[nums[i] - 1] == nums[i]) break; // can't just use nums[i] - 1 == i, deal with duplicates
        		swap(nums, i, nums[i] - 1);
        	}
        }
        for (int i = 0; i < nums.length; i++) {
        	if (nums[i] != i + 1) return i + 1;
        }
        return nums.length + 1;
    }

    private void swap(int[] nums, int i, int j) {
    	if (nums == null || nums.length == 0) return;
    	if (i < 0 || j < 0 || i >= nums.length || j >= nums.length) return;
    	int tmp = nums[i];
    	nums[i] = nums[j];
    	nums[j] = tmp;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {4, -2, 7, 1 ,3};
    	System.out.println(sol.firstMissingPositive(t));
    }
}